File: core/browseruse_engine/agent/message_manager.py
"""
BrowserUse Engine - 消息历史管理器

管理 LLM 对话历史，支持上下文压缩。
"""

import logging
from typing import List, Optional, Any
from dataclasses import dataclass, field

from ..llm.base import BaseMessage, SystemMessage, UserMessage, AssistantMessage

logger = logging.getLogger(__name__)


class MessageManager:
    """
    消息历史管理器

    负责维护 LLM 对话历史，支持：
    - 消息添加和获取
    - 历史截断
    - 上下文压缩
    """

    def __init__(
        self,
        max_messages: int = 50,
        max_tokens_estimate: int = 100000,
    ):
        """
        初始化消息管理器

        Args:
            max_messages: 最大消息数量
            max_tokens_estimate: 估计的最大 token 数
        """
        self.max_messages = max_messages
        self.max_tokens_estimate = max_tokens_estimate
        self._messages: List[BaseMessage] = []
        self._system_message: Optional[SystemMessage] = None

    def add_system_message(self, content: str) -> None:
        """
        设置系统消息

        系统消息始终位于对话开头。
        """
        self._system_message = SystemMessage(content)

    def add_user_message(
        self,
        content: str,
        image_base64: Optional[str] = None
    ) -> None:
        """
        添加用户消息

        Args:
            content: 消息内容
            image_base64: 图片 base64 编码 (可选)
        """
        if image_base64:
            message = UserMessage.with_image(content, image_base64)
        else:
            message = UserMessage(content)
        self._messages.append(message)
        self._trim_if_needed()

    def add_assistant_message(self, content: str) -> None:
        """添加助手消息"""
        self._messages.append(AssistantMessage(content))
        self._trim_if_needed()

    def get_messages(self) -> List[BaseMessage]:
        """
        获取完整消息列表

        Returns:
            消息列表 (系统消息 + 历史消息)
        """
        messages = []
        if self._system_message:
            messages.append(self._system_message)
        messages.extend(self._messages)
        return messages

    def get_last_n_messages(self, n: int) -> List[BaseMessage]:
        """获取最近 n 条消息"""
        messages = []
        if self._system_message:
            messages.append(self._system_message)
        messages.extend(self._messages[-n:])
        return messages

    def _trim_if_needed(self) -> None:
        """如果超过限制，裁剪消息"""
        if len(self._messages) > self.max_messages:
            # 保留最近的消息
            trim_count = len(self._messages) - self.max_messages
            self._messages = self._messages[trim_count:]
            logger.debug(f"裁剪了 {trim_count} 条消息")

    def clear(self) -> None:
        """清空消息历史 (保留系统消息)"""
        self._messages.clear()

    def compress_history(self, keep_recent: int = 10) -> str:
        """
        压缩历史记录

        将旧消息压缩为摘要，保留最近的消息。

        Args:
            keep_recent: 保留最近的消息数量

        Returns:
            压缩后的摘要
        """
        if len(self._messages) <= keep_recent:
            return ""

        # 提取要压缩的消息
        to_compress = self._messages[:-keep_recent]
        kept = self._messages[-keep_recent:]

        # 生成摘要 (简单版本)
        summary_parts = []
        for msg in to_compress:
            if isinstance(msg, UserMessage):
                content = msg.content if isinstance(msg.content, str) else str(msg.content)
                summary_parts.append(f"User: {content[:100]}...")
            elif isinstance(msg, AssistantMessage):
                content = msg.content if isinstance(msg.content, str) else str(msg.content)
                summary_parts.append(f"Assistant: {content[:100]}...")

        summary = "\n".join(summary_parts[-5:])  # 只保留最后 5 条摘要

        # 更新消息列表
        self._messages = kept

        return summary

    @property
    def message_count(self) -> int:
        """消息数量"""
        return len(self._messages)

    @property
    def has_system_message(self) -> bool:
        """是否有系统消息"""
        return self._system_message is not None

    def estimate_tokens(self) -> int:
        """
        估计当前消息的 token 数

        使用简单的字符数估计 (1 token ≈ 4 字符)
        """
        total_chars = 0
        if self._system_message:
            content = self._system_message.content
            if isinstance(content, str):
                total_chars += len(content)

        for msg in self._messages:
            content = msg.content
            if isinstance(content, str):
                total_chars += len(content)
            elif isinstance(content, list):
                for item in content:
                    if isinstance(item, dict) and "text" in item:
                        total_chars += len(item["text"])

        return total_chars // 4

File: desktop/src/browseruse/agent/message-manager.ts
/**
 * BrowserUse Engine - 消息历史管理器（Node 重写）
 * 对标 core/browseruse_engine/agent/message_manager.py
 *
 * 移植说明：
 *   - Python 的 SystemMessage/UserMessage/AssistantMessage 类实例化 →
 *     ../llm/base.ts 的 createSystemMessage / createUserMessage /
 *     createUserMessageWithImage / createAssistantMessage 工厂
 *   - Python 用 isinstance(msg, UserMessage) 区分消息类型；TS 侧没有类，
 *     改用 msg.role === "user" / "assistant" 判定（语义等价）
 *   - 构造参数 max_messages/max_tokens_estimate → 对象形式的 camelCase 选项，
 *     与 AgentService 的构造风格保持一致；默认值逐字照搬（50 / 100000）
 *   - logger.debug → 注入的 LogFn（默认 noopLog）
 *   - str(content) 在 Node 没有等价 repr，非字符串内容用 JSON.stringify
 *     （只影响 compressHistory 的摘要文本，不影响判定逻辑）
 *   - len() 按 Unicode 码点、String.length 按 UTF-16 码元，
 *     estimateTokens 的字符统计在星形平面字符上会略有差异（本身就是粗估）
 */

import type { BaseMessage } from "../llm/base.ts";
import {
  createAssistantMessage,
  createSystemMessage,
  createUserMessage,
  createUserMessageWithImage,
} from "../llm/base.ts";
import { noopLog, type LogFn } from "../page.ts";

export interface MessageManagerOptions {
  /** 最大消息数量 */
  maxMessages?: number;
  /** 估计的最大 token 数 */
  maxTokensEstimate?: number;
  log?: LogFn;
}

/**
 * 消息历史管理器
 *
 * 负责维护 LLM 对话历史，支持：
 * - 消息添加和获取
 * - 历史截断
 * - 上下文压缩
 */
export class MessageManager {
  readonly maxMessages: number;
  readonly maxTokensEstimate: number;
  private readonly log: LogFn;
  private _messages: BaseMessage[] = [];
  private _systemMessage: BaseMessage | null = null;

  constructor(options: MessageManagerOptions = {}) {
    this.maxMessages = options.maxMessages ?? 50;
    this.maxTokensEstimate = options.maxTokensEstimate ?? 100000;
    this.log = options.log ?? noopLog;
  }

  /** 设置系统消息（系统消息始终位于对话开头） */
  addSystemMessage(content: string): void {
    this._systemMessage = createSystemMessage(content);
  }

  /**
   * 添加用户消息
   *
   * @param content 消息内容
   * @param imageBase64 图片 base64 编码 (可选)
   */
  addUserMessage(content: string, imageBase64?: string | null): void {
    let message: BaseMessage;
    if (imageBase64) {
      message = createUserMessageWithImage(content, imageBase64);
    } else {
      message = createUserMessage(content);
    }
    this._messages.push(message);
    this._trimIfNeeded();
  }

  /** 添加助手消息 */
  addAssistantMessage(content: string): void {
    this._messages.push(createAssistantMessage(content));
    this._trimIfNeeded();
  }

  /** 获取完整消息列表（系统消息 + 历史消息） */
  getMessages(): BaseMessage[] {
    const messages: BaseMessage[] = [];
    if (this._systemMessage) {
      messages.push(this._systemMessage);
    }
    messages.push(...this._messages);
    return messages;
  }

  /** 获取最近 n 条消息 */
  getLastNMessages(n: number): BaseMessage[] {
    const messages: BaseMessage[] = [];
    if (this._systemMessage) {
      messages.push(this._systemMessage);
    }
    // Python 的 self._messages[-n:]：n=0 时取全量，slice(-0) 行为一致
    messages.push(...this._messages.slice(-n));
    return messages;
  }

  /** 如果超过限制，裁剪消息 */
  private _trimIfNeeded(): void {
    if (this._messages.length > this.maxMessages) {
      // 保留最近的消息
      const trimCount = this._messages.length - this.maxMessages;
      this._messages = this._messages.slice(trimCount);
      this.log(`裁剪了 ${trimCount} 条消息`);
    }
  }

  /** 清空消息历史 (保留系统消息) */
  clear(): void {
    this._messages = [];
  }

  /**
   * 压缩历史记录：将旧消息压缩为摘要，保留最近的消息。
   *
   * @param keepRecent 保留最近的消息数量
   * @returns 压缩后的摘要
   */
  compressHistory(keepRecent = 10): string {
    if (this._messages.length <= keepRecent) {
      return "";
    }

    // 提取要压缩的消息
    const toCompress = this._messages.slice(0, this._messages.length - keepRecent);
    const kept = this._messages.slice(-keepRecent);

    // 生成摘要 (简单版本)
    const summaryParts: string[] = [];
    for (const msg of toCompress) {
      if (msg.role === "user") {
        const content = typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content);
        summaryParts.push(`User: ${content.slice(0, 100)}...`);
      } else if (msg.role === "assistant") {
        const content = typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content);
        summaryParts.push(`Assistant: ${content.slice(0, 100)}...`);
      }
    }

    const summary = summaryParts.slice(-5).join("\n"); // 只保留最后 5 条摘要

    // 更新消息列表
    this._messages = kept;

    return summary;
  }

  /** 消息数量 */
  get messageCount(): number {
    return this._messages.length;
  }

  /** 是否有系统消息 */
  get hasSystemMessage(): boolean {
    return this._systemMessage !== null;
  }

  /**
   * 估计当前消息的 token 数
   *
   * 使用简单的字符数估计 (1 token ≈ 4 字符)
   */
  estimateTokens(): number {
    let totalChars = 0;
    if (this._systemMessage) {
      const content = this._systemMessage.content;
      if (typeof content === "string") {
        totalChars += content.length;
      }
    }

    for (const msg of this._messages) {
      const content = msg.content;
      if (typeof content === "string") {
        totalChars += content.length;
      } else if (Array.isArray(content)) {
        for (const item of content) {
          if (item !== null && typeof item === "object" && "text" in item) {
            const text = (item as { text?: unknown }).text;
            if (typeof text === "string") totalChars += text.length;
          }
        }
      }
    }

    return Math.floor(totalChars / 4);
  }
}
